Introduce a framework for testing serialization and hashing across sessions

dolio opened the pull request:
This introduces some stuff to `base.u` for serializing test cases to files and loading them back up later. The test files contain all the code necessary to run the serialized test value(s), so values can be saved in one transcript file, then loaded in a separate transcript file, to test how things would actually happen when transmitting values across machines.

So far I've written one transcript to generate a serialized test case, and the transcript that loads and runs all the serialized test cases in a random order. I'll write some additional generator transcripts after submitting this draft. Both the generators and the random test would be run on every CI, presumably, which allows checking that the generators produce the same test cases (serialized output and hash) and that the serialized test cases work when loaded in a random order.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/unisonweb/unison/3401)
<!-- Reviewable:end -->

dolio committed:
Transcript base additions for serialized test cases

dolio committed:
Add hashes to serial test cases

- With first test case generator transcript

dolio committed:
Avoid saving code for builtins in base.u

- Getting the wrapper code for builtins was added for the compiler, but
  deserialization isn't supported for foreign ops, so just avoid it.

dolio committed:
Second test case

dolio committed:
Fix randomization in random-deserial test

dolio committed:
Update 00 test case for absence of serialized builtins

dolio committed:
Use group reference when emitting machine code

- Previously, a dummy reference ##Continuation was used when emitting a
  let that had to call into a separate definition. This didn't matter
  locally, because the references are just used for debugging
  information. However, when serializing, the reference is used because
  it is externally stable. This meant that continuations with let
  resumptions couldn't be decompiled successfully, because they were all
  referring to the same ##Continuation reference that didn't map to any
  actual combinator.
- Instead, we use the reference for the combinator being compiled, since
  it will have the code that is actually needed.

dolio committed:
More test cases

dolio committed:
random-deserial: sort test case results for consistent output

dolio commented:
@aryairani or @ChrisPenner Will the files in `serialized-cases` automatically get checked for differences by CI, because it's nested under `unison-src/transcripts-using-base`? Or do I have to set something extra up for that?

ChrisPenner commented:
@dolio The way the transcript testing works is that one of the CI steps literally runs this:

```
stack --no-terminal exec transcripts
# Add all changes to the index for when we diff.
git add --all
# Fail if any transcripts cause git diffs.
git diff --cached --ignore-cr-at-eol --exit-code
```

So if `stack exec transcripts` would make changes tracked by git, then CI will catch it, otherwise not.

dolio commented:
Okay, cool. I think that means it will catch things the way I want.

dolio committed:
Store hashes in base32 for nicer display

dolio committed:
Fix tests

dolio committed:
Allow serializing nested code/value terms

- Code containing foreign operations cannot be successfully serialized
  this way. Deserializing such things is already unsupported, however.
  Right now such serialization is only used for the scheme compiler.

dolio committed:
Merge branch 'trunk' into work/serial-tests

dolio commented:
Wasn't sure who to put for review, so I just checked a few people who might be interested.

Addresses #3379

ChrisPenner commented on unison-src/transcripts-using-base/base.u:
```diff
@@ -224,3 +234,133 @@ MVar.take = compose reraise take.impl
+  [] -> found
+  ln +: rest
+    | Text.take 2 (Link.Term.toText ln) == "##" -> crawl found rest
+    | bContains found ln -> crawl found rest
```
bContains requires a sorted list right?

Is `found` guaranteed to be sorted?
Also would likely be worth mentioning the sorting invariants that are required on each of the functions, e.g. `bContains`, `bSearch`

EDIT:
I see that `found` is built up using `bInsert`, but since it could still be called externally with a list from elsewhere it'd probably still be good to mention the invariant 😄

dolio replied:
Okay, I added some comments to these.

ChrisPenner commented on unison-src/transcripts-using-base/random-deserial.md:
```diff
@@ -0,0 +1,57 @@
+
+gen : Nat -> Nat -> (Nat, Nat)
+gen seed k =
+  c = 1442695040888963407
```
Chosen by fair dice roll, guaranteed to be random? 😋

dolio replied:
I just copied it out of base. No idea how Paul (I think) chose it. :)

ChrisPenner reviewed (approved):
Admittedly I don't fully understand the whole implementation, but I'm definitely glad we're testing this!

dolio committed:
Add a bit of documentation to base.u

dolio committed:
More base.u documentation

stew reviewed (approved):
the description of what this does makes total sense to me as to why we are doing what we are doing, no idea if the implementation does this :)

ceedubs reviewed (approved):
ditto what Stew said :)

ceedubs commented:
@dolio it looks like builds have been failing since this was merged. Was there a bad merge or something? https://github.com/unisonweb/unison/actions/runs/3085370292/jobs/4988583683#step:15:4382
